### ISSUE TITLE
standardize commandline patterns

### ISSUE THREAD

**Issue by dwiel** [knausj85/knausj_talon#231]
There are many ways to create grammars for command line programs. I will enumerate a few options using git as an example:

A) no new lines ever
```
git push: "git push "
```
B) new lines sometimes (perhaps I when I use "git diff" I'm always comparing branches, never using it on its own for example)
```
git push: "git push\n"
git pull rebase: "git pull --rebase\n"
git move: "git mv "
git diff: "git diff"
```
C) new lines everywhere (except where it obviously doesn't make sense)
```
git push: "git push\n"
git pull rebase: "git pull --rebase\n"
git move: "git mv "
git diff: "git diff\n"
```

and regarding text captures a similar design decision

X) no text captures anywhere
```
git commit message: "git commit -m '[|]'"
git check out: "git checkout "
```
Y) text captures on incantations which often get followed by additional text (perhaps I almost never use any additional arguments to "git pull")
```
git commit message <user.optional_text>: "git commit -m '{optional_text}'"
git pull: "git pull"
```
Z) text captures anywhere free text it is valid
```
git add <user.optional_text>: "git add {user.optional_text}"
git push force <user.optional_text>: "git push --force {user.optional_text}"
```

Many of these patterns are applicable to lots of commandline programs. it would be nice to understand the trade-offs of different styles and select a single style and keep everything consistent and predictable. Perhaps there are additional styles I'm not including here, feel free to add to the list.

**Comment by dwiel** [comment]
A

**Comment by dwiel** [comment]
B

**Comment by dwiel** [comment]
C

**Comment by dwiel** [comment]
X

**Comment by dwiel** [comment]
Y

**Comment by dwiel** [comment]
Z

**Comment by rntz** [comment]
In many cases you don't want a full text capture, only a word capture at most. For example, a branch name is never more than a single word, although you might want to be able to use a formatter. For example, I have:

```
git fetch: "git fetch "
git fetch <user.word>: "git fetch {word}"
git merge: "git merge "
git merge <user.word>: "git merge {word}"
... etc ...
```

I tend to prefer no new lines by default, because you can always say "enter" afterwards, but there's nothing you can say to prevent a new line if the command adds one. Although another approach would be to do exactly that:

```
git merge: "git merge\n"
git merge and: "git merge "
git pull: "git pull\n"
git pull and: "git pull "
```

**Comment by rntz** [comment]
D) New lines by default, with a suffix to disable

**Comment by fidgetingbits** [comment]
I don't know how often people will be chaining command line commands together such that the difference between word or text capture is noticeably different? but I guess that raises the question (for me anyways), with a word matching versus text caption matching actually result in something significantly different or is it effectively the same with text matching just allowing you to say even more words where necessary? if they're effectively the same then I guess if it's a text capture, in the ability to chain these types of commands is important,  it could sometimes be a text capture that has `over` baked in to end the capture?

**Comment by dwiel** [comment]
Yeah baking in over seems like a reasonable approach to make it easier/more explicit when chaining commands.

The main issue for me around all of this is just minimizing cognitive overhead in remembering which commands are set up in which way, especially commands I may only use once every month or two.

**Comment by rntz** [comment]
The lowest cognitive overhead is probably to never have new lines **or** text captures. You can always use `enter`/`phrase|say|speak`.  This is also easiest to implement, and makes it obvious how to handle other command line programs. I would be fine with this. It is somewhat inconvenient, which is why I prefer having text captures for commands that nearly always need them, like `git commit`.

As for word versus text capture, I chain commands a lot, but a lot of the time it's just `enter/slap` to run the git command :P. Making the captures non-eager is probably enough.

When you mention baking in over, what do you mean? Do you mean using non-eager captures, or having two variants, one of which uses over?

Non-eager captures:
```
git merge [<user.text>]: ...
```

Both variants:
```
git merge: ...
git merge <user.text>$: ...
git merge <user.text> over: ...
```

If we're going to have text captures than I prefer non-eager captures, because they're better for chaining. But having both variants is more consistent with the way knausj `phrase|say|speak` works.

**Comment by dwiel** [comment]
Random idea: what if we use a capture like `<user.command_line_text>` which can then be easily changed in one place: `[<user.text>]` or `<user.text> (over|$)` or `[never match anything]` etc.

Maybe its overkill, just an idea.

**Comment by dwiel** [comment]
I also prefer non-eager captures, but with a user.command_line_text I think we could easily make that configurable.

I guess baking in over to me meant using non eager capture `[<user.text> [over]]`

**Comment by rntz** [comment]
A user-configurable capture would work, but it seems like the kind of over-abstraction that aegis would be unhappy about.

I don't think there's much point in using both a non-eager capture **and** over, like so:
```
git merge [<user.text> [over]]: ...
```
AIUI the whole point of `over` is that it allows you to also have an eager anchored version; otherwise `over` is redundant. You might as well omit it and have a no-op command `over:`
```
git merge [<user.text>]: ...
over: skip()
```
(In fact I have such an `over` command, although I don't think I really need it.)

**Comment by dwiel** [comment]
Yeah it might be over abstracting, but I also think merge conflicts wind up resulting in fewer people willing to contribute back to central repos and pulling updates from central repos. Maybe we should try to pick one method and abstract if it looks like that would help a specific problem.

**Comment by dwiel** [comment]
Over is still helpful in non-eager mode to make sure text ends up in the capture and not as a command. For example: `git commit message fix bug caused by user pressing enter` will actually press enter at the end even if you wanted the text "enter". Similar for text ending in made, for, sit, end, home, etc.

**Comment by dwiel** [comment]
okay, so it looks like the most popular option is:

```
git commit message <user.optional_text>: 'git commit -m "{optional_text}[|]"'
git pull <user.optional_text>: 'git pull {optional_text}'
```

Another proposed option is:

```
git pull <word>: "git pull {word}"
```

though this is a bit awkward in the case of `git pull origin master`.

Another proposed (and currently implemented in master) option is:

```
git pull: "git pull\n"
git pull <user.text>: "git pull {text} "
```

I'm not sure how to reconcile these differences. Any ideas? Are these two the primary desired patterns or am I reading the thread wrong? Just trying to summarize so we can come to a conclusion :-)
